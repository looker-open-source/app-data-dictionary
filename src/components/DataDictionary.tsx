/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import React, { PureComponent, useState } from "react";
import {
  Chip,
  Flex,
  FlexItem,
  Heading,
  Spinner,
  theme,
} from "@looker/components";
import humanize from 'humanize-string'
import styled, { ThemeProvider } from "styled-components";
import { useAllModels } from "../utils/fetchers";
import "./styles.css";
import { PanelFields } from "./PanelFields";
import SidebarToggle from "./SidebarToggle";
import { useCurrentModel, useCurrentExplore } from "../utils/routes"
import { ColumnDescriptor } from "./interfaces";
import { SQLSnippet } from "./SQLSnippet";
import { Sidebar } from './Sidebar'
import { SidebarStyleProps } from "./interfaces";
import { NoModelsAvailable } from "./NoModelsAvailable";
import { ILookmlModelExploreField } from "@looker/sdk";
import { CategorizedLabel } from './CategorizedLabel'

export const columns: ColumnDescriptor[] = [
  {
    name: 'field-label',
    label: 'Field Label',
    rowValueDescriptor: 'label_short',
    formatter: (x: any, isRow: boolean, field: ILookmlModelExploreField) => {
      if (field.field_group_label && field.field_group_variant) {
        return `${field.field_group_label} ${field.field_group_variant}`
      } else {
        return x
      }
    },
    minWidth: '12em',
  },
  {
    name: 'category',
    label: 'Category',
    rowValueDescriptor: 'category',
    formatter: (x: any, isRow: boolean, field: ILookmlModelExploreField) => {
      return <CategorizedLabel label={x} category={field.category} />
    },
    minWidth: '12em',
  },
  {
    name: 'description',
    label: 'Description',
    rowValueDescriptor: 'description',
    formatter: (x: any, isRow: boolean) => {
      if (x && isRow && x.length > 200) {
        return x.substring(0, 200) + '...'
      }
      return x
    },
    maxWidth: '20em',
  },
  {
    name: 'lookml-name',
    label: 'LookML Name',
    rowValueDescriptor: 'name',
    formatter: (x: any) => {
      return x.replace(/\./g, '.\u200B');
    },
    minWidth: '15em',
  },
  {
    name: 'type',
    label: 'Type',
    rowValueDescriptor: 'type',
    formatter: (x: any) => humanize(x),
    minWidth: '8em',
  },
  {
    label: 'SQL',
    rowValueDescriptor: 'sql',
    formatter: (x: any, isRow: boolean) => {
      return (<SQLSnippet isRow={isRow} src={x} />)
    },
    maxWidth: '20em',
    name: 'sql',
  },
  {
    name: 'tags',
    label: 'Tags',
    rowValueDescriptor: 'tags',
    formatter: (tags: any) => {
      return tags.map((tag: string) => (
        <Chip disabled>{tag}</Chip>
      ))
    },
  },
]

export const DataDictionary: React.FC<{}> = () => {
  const unfilteredModels = useAllModels()
  const currentModel = useCurrentModel()
  const [sidebarOpen, setSidebarOpen] = React.useState(true)
  const [search, setSearch] = React.useState('')
  const { currentExplore, loadingExplore } = useCurrentExplore()

  let models

  if (unfilteredModels) {
    models = unfilteredModels.filter(m => m.explores && m.explores.some(e => !e.hidden))
    if (!models.length) {
      return <NoModelsAvailable />
    }
  }

  if (!models) {
    return <Flex alignItems="center" height="100%" justifyContent="center"><Spinner /></Flex>
  }

  const toggleFn = () => setSidebarOpen(!sidebarOpen);

  return (
    <ThemeProvider theme={theme}>
      <div style={{minWidth: "1200px"}}>
        <PageHeader>
          <FlexItem>
            <Heading as="h1" fontSize="xlarge" fontWeight="semiBold" mb="xsmall">
              Data Dictionary
            </Heading>
          </FlexItem>
        </PageHeader>
        <PageLayout open={sidebarOpen}>
          <LayoutSidebar>
            {sidebarOpen && <Sidebar
              currentExplore={currentExplore}
              currentModel={currentModel}
              loadingExplore={loadingExplore}
              models={models}
              search={search}
              setSearch={setSearch}
            />}
          </LayoutSidebar>
          <SidebarDivider open={sidebarOpen}>
            <SidebarToggle
              isOpen={sidebarOpen}
              onClick={toggleFn}
              headerHeight="114px"
            />
          </SidebarDivider>
          <PageContent>
            <PanelFields
              columns={columns}
              currentExplore={currentExplore}
              currentModel={currentModel}
              loadingExplore={loadingExplore}
              model={currentModel}
            />
          </PageContent>
        </PageLayout>
      </div>
    </ThemeProvider>
  );
}


const PageHeader = styled(Flex)`
  background-color: ${theme.colors.palette.purple400};
  background-position: 100% 0;
  background-repeat: no-repeat;
  background-size: 836px 120px;
  padding: ${theme.space.large};
  background-image: url('https://marketplace-api.looker.com/misc/INC-100373/spirals.png');

  h1 {
    color: ${theme.colors.palette.white};
  }
`;

const PageLayout = styled.div<SidebarStyleProps>`
  display: grid;
  grid-template-rows: 1fr;
  grid-template-columns: ${({ open }) =>
    open ? "16.625rem 0 1fr" : "1.5rem 0 1fr"};
  grid-template-areas: "sidebar divider main";
  position: relative;
`;

const PageContent = styled.div`
  grid-area: main;
  position: relative;
`;

const LayoutSidebar = styled.aside`
  position: absolute;
  top: 0;
  bottom: 0;
  width: 16.625rem;
  grid-area: sidebar;
  z-index: 0;
`;

const SidebarDivider = styled.div<SidebarStyleProps>`
  transition: border 0.3s;
  border-left: 1px solid
    ${({ theme, open }) =>
      open ? theme.colors.palette.charcoal200 : "transparent"};
  grid-area: divider;
  overflow: visible;
  position: relative;
  &:hover {
    border-left: 1px solid
      ${({ theme, open }) =>
        open
          ? theme.colors.palette.charcoal300
          : theme.colors.palette.charcoal200};
  }
`;
