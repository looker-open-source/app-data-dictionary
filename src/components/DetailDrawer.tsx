/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import React from 'react'
import { theme, Drawer, TableRow, TableDataCell } from '@looker/components'
import styled from 'styled-components'
import type {
  ILookmlModel,
  ILookmlModelExplore,
  ILookmlModelExploreField,
  IUser,
} from '@looker/sdk'
import { useHistory } from 'react-router'
import { DETAILS_PANE, COMMENTS_PANE } from '../utils/constants'
import { internalExploreURL, usePathNames } from '../utils/routes'
import type {
  ColumnDescriptor,
  FieldComments,
  CommentPermissions,
} from './interfaces'
import { FieldMetadata } from './FieldMetadata'

const TableRowCustom = styled(TableRow as any)`
  transition: background-color 0.3s ease;
  &.active,
  &:hover {
    background-color: ${theme.colors.ui1};
  }

  .disabled {
    visibility: hidden;
  }

  &:hover .disabled {
    visibility: visible;
  }

  td {
    word-break: break-word;
  }
  cursor: pointer;
`

export const DetailDrawer: React.FC<{
  columns: ColumnDescriptor[]
  explore: ILookmlModelExplore
  model: ILookmlModel
  field: ILookmlModelExploreField
  shownColumns: string[]
  tab: number
  setTab: (tabIndex: number) => void
  comments: string
  addComment: (newCommentStr: string, field: string) => void
  editComment: (newCommentStr: string, field: string) => void
  deleteComment: (newCommentStr: string, field: string) => void
  authors: IUser[]
  me: IUser
  permissions: CommentPermissions
}> = ({
  columns,
  explore,
  field,
  model,
  shownColumns,
  tab,
  setTab,
  comments,
  addComment,
  editComment,
  deleteComment,
  authors,
  me,
  permissions,
}) => {
  const history = useHistory()
  const path = usePathNames()

  const parsedComments = JSON.parse(comments)

  const getFieldCommentsLength = (field: string) => {
    const exploreComments = parsedComments[explore.name]
      ? parsedComments[explore.name]
      : {}
    const commentFields = Object.keys(exploreComments)
    if (commentFields.includes(field) && exploreComments[field].length > 0) {
      return exploreComments[field].length
    } else {
      return null
    }
  }

  const fieldComments =
    (parsedComments[explore.name] &&
      parsedComments[explore.name][field.name]) ||
    []
  const sortedComments = fieldComments.sort(
    (x: FieldComments, y: FieldComments) => {
      return x.timestamp - y.timestamp
    }
  )

  const canViewComments = () => {
    if (permissions.disabled) {
      return false
    } else {
      return (
        (permissions.reader && sortedComments.length > 0) ||
        permissions.writer ||
        permissions.manager
      )
    }
  }

  function paneUrl(pane: number) {
    field &&
      history.push(
        internalExploreURL({
          model: explore.model_name,
          explore: explore.name,
          field: field.name,
          tab: pane.toString(),
        })
      )
  }

  function closePaneUrl() {
    field &&
      history.push(
        internalExploreURL({
          model: explore.model_name,
          explore: explore.name,
        })
      )
  }

  function detailsPane() {
    paneUrl(DETAILS_PANE)
    setTab(DETAILS_PANE)
  }

  function commentsPane() {
    canViewComments() && paneUrl(COMMENTS_PANE)
    canViewComments() && setTab(COMMENTS_PANE)
  }

  return (
    <Drawer
      content={
        <FieldMetadata
          field={field}
          columns={columns}
          explore={explore}
          key={field.name}
          model={model}
          tab={path.detailPane ? parseInt(path.detailPane) : tab}
          detailsPane={detailsPane}
          commentsPane={commentsPane}
          sortedComments={sortedComments}
          addComment={addComment}
          editComment={editComment}
          deleteComment={deleteComment}
          fieldCommentLength={getFieldCommentsLength(field.name)}
          commentAuthors={authors}
          me={me}
          permissions={permissions}
          canViewComments={canViewComments()}
        />
      }
      isOpen={field.name === path.fieldName}
      onClose={closePaneUrl}
    >
      <TableRowCustom>
        {columns.map(({ Formatter, ...props }) => {
          if (shownColumns.includes(props.rowValueDescriptor)) {
            const unformattedValue =
              field[props.rowValueDescriptor as keyof ILookmlModelExploreField]
            const displayValue = unformattedValue
              ? String(unformattedValue)
              : undefined
            return (
              <TableDataCell
                color="text3"
                p="medium"
                pl="small"
                fontWeight="normal"
                key={props.rowValueDescriptor}
                maxWidth={props.maxWidth}
                minWidth={props.minWidth}
                onClick={
                  props.rowValueDescriptor === 'comment'
                    ? commentsPane
                    : detailsPane
                }
              >
                <Formatter
                  x={displayValue}
                  isRow={true}
                  field={field}
                  commentCount={getFieldCommentsLength(field.name)}
                  canComment={canViewComments()}
                  reader={permissions.reader}
                  tags={unformattedValue as string[]}
                />
              </TableDataCell>
            )
          }
          return undefined
        })}
      </TableRowCustom>
    </Drawer>
  )
}
