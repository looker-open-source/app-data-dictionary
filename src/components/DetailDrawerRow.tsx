/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import React from 'react'
import { Code, TableRow, TableDataCell } from '@looker/components'

import type { ILookmlModelExploreField } from '@looker/sdk'
import type { ColumnDescriptor } from './interfaces'

export const DetailDrawerRow: React.FC<{
  column: ColumnDescriptor
  field: ILookmlModelExploreField
}> = ({ column, field }) => {
  const unformattedValue =
    field[column.rowValueDescriptor as keyof ILookmlModelExploreField]
  const displayValue = unformattedValue ? String(unformattedValue) : undefined
  const Formatter = column.Formatter
  return (
    <TableRow key={column.rowValueDescriptor}>
      <TableDataCell color="text" p="medium" pl="small" pr="small">
        {column.label}
      </TableDataCell>
      <TableDataCell
        className="break"
        color="text3"
        p="medium"
        pl="small"
        pr="small"
      >
        <Code color="text3" fontSize="small" className="break">
          <Formatter
            x={displayValue}
            isRow={false}
            field={field}
            tags={unformattedValue as string[]}
          />
        </Code>
      </TableDataCell>
    </TableRow>
  )
}
