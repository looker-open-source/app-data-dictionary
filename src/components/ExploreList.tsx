/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import React from 'react'
import { FlexItem, List, ListItem, theme } from '@looker/components'
import './styles.css'
import { useHistory } from 'react-router'
import type { ILookmlModelExplore } from '@looker/sdk'
import { internalExploreURL, useCurrentModel } from '../utils/routes'

const isActive = (
  currentExplore: ILookmlModelExplore,
  listExplore: ILookmlModelExplore,
  loadingExplore: string
) => {
  if (
    !loadingExplore &&
    currentExplore &&
    listExplore.name === currentExplore.name
  ) {
    return 'active'
  } else if (loadingExplore && listExplore.name === loadingExplore) {
    return 'active'
  } else {
    return null
  }
}

export const ExploreList: React.FC<{
  currentExplore: ILookmlModelExplore
  search: string
  loadingExplore: string
}> = ({ currentExplore, loadingExplore, search }) => {
  const currentModel = useCurrentModel()
  const history = useHistory()
  return (
    <FlexItem pt="xlarge" ml="xsmall">
      <List density={0} color={theme.colors.key}>
        {currentModel &&
          currentModel.explores &&
          currentModel.explores.map((explore: any) => {
            if (
              !explore.hidden &&
              (!search ||
                explore.label.toLowerCase().includes(search.toLowerCase()))
            ) {
              return (
                <ListItem
                  key={explore.name}
                  truncate={true}
                  onClick={() => {
                    history.push(
                      internalExploreURL({
                        model: currentModel.name || '',
                        explore: explore.name,
                      })
                    )
                  }}
                  selected={isActive(currentExplore, explore, loadingExplore)}
                >
                  {explore.label}
                </ListItem>
              )
            }
            return undefined
          })}
      </List>
    </FlexItem>
  )
}
