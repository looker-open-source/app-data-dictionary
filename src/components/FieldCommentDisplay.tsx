/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import React from 'react'
import {
  FlexItem,
  Card,
  CardContent,
  Flex,
  AvatarUser,
  Text,
  IconButton,
  SpaceVertical,
  Menu,
  MenuItem,
} from '@looker/components'
import { MoreVert, Create, Delete } from '@styled-icons/material'
import styled from 'styled-components'
import type { FieldComments, UserData } from './interfaces'

const CustomCommentCard = styled(Card as any)`
  .show {
    visibility: hidden;
  }

  .hide {
    visibility: hidden;
  }

  &:hover .show {
    visibility: visible;
  }
`

export const FieldCommentDisplay: React.FC<{
  authorData: UserData
  comment: FieldComments
  showDetails: () => string
  toggleEdit: () => void
  openDialog: () => void
}> = ({ authorData, comment, showDetails, toggleEdit, openDialog }) => {
  const timestamp = new Date(comment.timestamp)

  return (
    <CustomCommentCard aria-label="FieldCommentDisplay">
      <CardContent>
        <FlexItem>
          <Flex pb="small">
            <FlexItem flexBasis="10%" pr="xsmall" pt="xxxsmall">
              <AvatarUser user={authorData} size="xsmall" />
            </FlexItem>
            <FlexItem flexBasis="80%" aria-label="FieldCommentDisplayNameTime">
              <FlexItem>
                <Text fontSize="small" fontWeight="semiBold">
                  {authorData.display_name}
                </Text>
              </FlexItem>
              <FlexItem>
                <Text fontSize="xsmall" variant="secondary">
                  {timestamp.toLocaleString()}
                  {comment.edited ? ' (edited)' : null}
                </Text>
              </FlexItem>
            </FlexItem>
            <FlexItem flexBasis="10%">
              <SpaceVertical align="end">
                <Menu
                  density={-1}
                  content={
                    <>
                      <MenuItem onClick={toggleEdit} icon={<Create />}>
                        Edit Comment
                      </MenuItem>
                      <MenuItem onClick={openDialog} icon={<Delete />}>
                        Delete Comment
                      </MenuItem>
                    </>
                  }
                >
                  <IconButton
                    icon={<MoreVert />}
                    label="More Options"
                    className={showDetails()}
                  />
                </Menu>
              </SpaceVertical>
            </FlexItem>
          </Flex>
          <Text fontSize="small" aria-label="FieldCommentDisplayContent">
            {decodeURI(comment.content)}
          </Text>
        </FlexItem>
      </CardContent>
    </CustomCommentCard>
  )
}
