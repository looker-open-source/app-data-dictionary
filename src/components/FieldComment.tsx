/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import type { SyntheticEvent } from 'react'
import React from 'react'
import {
  ButtonOutline,
  Button,
  FlexItem,
  FieldTextArea,
  useConfirm,
  Space,
} from '@looker/components'

import type { IUser, ILookmlModelExploreField } from '@looker/sdk'
import { NOT_EDITING_COMMENT } from '../utils/constants'
import type { FieldComments, UserData, CommentPermissions } from './interfaces'
import { FieldCommentDisplay } from './FieldCommentDisplay'

export const FieldComment: React.FC<{
  comment: FieldComments
  editingComment: string
  setEditingComment: (editingCommentPk: string) => void
  commentContent: string
  setCommentContent: (newCommentContent: string) => void
  editComment: (newCommentStr: string, field: string) => void
  deleteComment: (newCommentStr: string, field: string) => void
  field: ILookmlModelExploreField
  authorData: UserData
  me: IUser
  addingNew: boolean
  permissions: CommentPermissions
}> = ({
  comment,
  editingComment,
  setEditingComment,
  setCommentContent,
  editComment,
  deleteComment,
  field,
  commentContent,
  authorData,
  me,
  addingNew,
  permissions,
}) => {
  const showDetails = () => {
    if (permissions.manager || (comment.author === me.id && !addingNew)) {
      return 'show'
    }
    return 'hide'
  }

  const toggleEdit = () => {
    editingComment === NOT_EDITING_COMMENT
      ? setEditingComment(comment.pk)
      : setEditingComment(NOT_EDITING_COMMENT)
  }

  const endEdit = () => {
    setEditingComment(NOT_EDITING_COMMENT)
  }

  const handleChange = (e: SyntheticEvent) => {
    const target = e.target as HTMLTextAreaElement
    setCommentContent(encodeURI(target.value))
  }

  const addToComments = () => {
    const newComment = {
      author: comment.author,
      edited: true,
      timestamp: comment.timestamp,
      content: commentContent,
      pk: comment.pk,
    }
    editComment(JSON.stringify(newComment), field.name)
    toggleEdit()
  }

  function deleteFromComments(close: any) {
    comment.deleted = true
    deleteComment(JSON.stringify(comment), field.name)
    close()
  }

  const [confirmationDialog, openDialog] = useConfirm({
    confirmLabel: 'Delete',
    buttonColor: 'critical',
    title: `Delete Comment?`,
    message:
      'Deleting this comment will permanently remove it. You cannot undo this later.',
    onConfirm: deleteFromComments,
  })

  return (
    <FlexItem>
      {confirmationDialog}
      {editingComment && comment.pk === editingComment ? (
        <FlexItem>
          <FieldTextArea
            autoFocus
            onChange={handleChange}
            defaultValue={decodeURI(comment.content)}
          />
          <Space pt="small" gap="xsmall" reverse>
            <Button size="medium" onClick={addToComments}>
              Save
            </Button>
            <ButtonOutline size="medium" color="neutral" onClick={endEdit}>
              Cancel
            </ButtonOutline>
          </Space>
        </FlexItem>
      ) : (
        <FieldCommentDisplay
          authorData={authorData}
          comment={comment}
          showDetails={showDetails}
          toggleEdit={toggleEdit}
          openDialog={openDialog}
        />
      )}
    </FlexItem>
  )
}
