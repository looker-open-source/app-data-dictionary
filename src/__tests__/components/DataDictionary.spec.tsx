/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import React from 'react'
import { screen } from '@testing-library/react'
import { renderWithExtensionContext } from '../test_utils/render_with_extension'
import {
  mockModels,
  mockComments,
  mockCurrentModel,
  mockCurrentExplore,
} from '../MockData/MockData'
import { DataDictionary } from '../../components/DataDictionary'

jest.mock('../../utils/fetchers', () => {
  return {
    useAllModels: jest.fn(() => {
      return mockModels
    }),
    useComments: jest.fn(() => {
      return mockComments
    }),
  }
})

jest.mock('../../utils/routes', () => {
  return {
    useCurrentModel: jest.fn(() => {
      return mockCurrentModel
    }),
    useCurrentExplore: jest.fn(() => {
      return mockCurrentExplore
    }),
  }
})

jest.mock('../../components/PanelFields', () => ({
  PanelFields: () => 'PanelFields',
}))

jest.mock('../../components/Sidebar', () => ({
  Sidebar: () => 'Sidebar',
}))

jest.mock('../../components/CategorizedLabel', () => ({
  CategorizedLabel: () => 'CategorizedLabel',
}))

describe('<DataDictionary>', () => {
  it('renders correctly', () => {
    renderWithExtensionContext(<DataDictionary />)
    expect(screen.getByText(/Data Dictionary/)).toBeInTheDocument()
    expect(screen.getByText(/Close Sidebar/)).toBeInTheDocument()
    expect(screen.getByText(/PanelFields/)).toBeInTheDocument()
  })
})
