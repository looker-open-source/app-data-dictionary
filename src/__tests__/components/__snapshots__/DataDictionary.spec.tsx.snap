// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  style={
    Object {
      "minWidth": "1200px",
    }
  }
>
  <div
    className="Flex-sc-vm49hu sc-EHOje cqXEQA"
  >
    <div
      className="FlexItem-sc-1re6mwg eLXSKG"
    >
      <h1
        className="TextBase-sc-1rkc9wq Heading-sc-cg3kf8 cbeLsS"
        fontSize="xlarge"
        fontWeight="semiBold"
      >
        Data Dictionary
      </h1>
    </div>
  </div>
  <div
    className="sc-bZQynM daVCun"
    open={true}
  >
    <aside
      className="sc-htoDjs cybUkI"
    >
      Sidebar
    </aside>
    <div
      className="sc-dnqmqq geodzx"
      open={true}
    >
      <div
        className="sc-bdVaJa fENiVh"
      >
        <button
          className="ButtonOuter-sc-wugisi eFTclI ButtonBase-sc-18wmtoc ButtonTransparent-sc-1gse89j cYulav IconButton-sc-1yur21t cnuNGW"
          color="neutral"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyUp={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
          shape="round"
          size="small"
          type="button"
          width={28}
        >
          <div
            className="VisuallyHidden-sc-10j76d8 gsEqg"
          >
            Close Sidebar
          </div>
          <div
            aria-hidden={true}
            className="Styled-sc-1rq5u7z eqQSq Icon-sc-1079d6u cWvDzp"
            size={22}
          >
            <svg
              fill="currentColor"
              height="100%"
              viewBox="0 0 24 24"
              width="100%"
            >
              <title>
                Caret Left
              </title>
              <path
                d="M15.41 16.59L10.83 12l4.58-4.59L14 6l-6 6 6 6 1.41-1.41z"
                fill="currentColor"
              />
            </svg>
          </div>
        </button>
      </div>
    </div>
    <div
      className="sc-gzVnrw feoOem"
    >
      PanelFields
    </div>
  </div>
</div>
`;
